app/scripts/controllers/permissions/index.js

```diff
@@ -0,0 +1,418 @@
+   */
+  _requestPermissions (origin, permissions) {
+    return new Promise((resolve, reject) => {
+
+      const req = { method: 'wallet_requestPermissions', params: [permissions] }
+      const res = {}
+      this.permissions.providerMiddlewareFunction(
+        { origin }, req, res, () => {}, _end
+      )
+
+      function _end (err) {
+        if (err || res.error) {
+          reject(err || res.error)
+        } else {
+          resolve(res.result)
+        }
+      }
+    })
+  }
+
+  /**
+   * User approval callback. The request can fail if the request is invalid.
+   *
+   * @param {object} approved the approved request object
+   */
+  async approvePermissionsRequest (approved, accounts) {
+
+    const { id } = approved.metadata
+    const approval = this.pendingApprovals[id]
+
+    try {
+
+      // attempt to finalize the request and resolve it
+      await this.finalizePermissionsRequest(approved.permissions, accounts)
+      approval.resolve(approved.permissions)
+
+    } catch (err) {
+
+      // if finalization fails, reject the request
+      approval.reject(ethErrors.rpc.invalidRequest({
+        message: err.message, data: err,
+      }))
+    }
+
+    delete this.pendingApprovals[id]
+  }
+
+  /**
+   * User rejection callback.
+   *
+   * @param {string} id the id of the rejected request
+   */
+  async rejectPermissionsRequest (id) {
+    const approval = this.pendingApprovals[id]
+    approval.reject(ethErrors.provider.userRejectedRequest())
+    delete this.pendingApprovals[id]
+  }
+
+  /**
+   * Grants the given origin the eth_accounts permission for the given account(s).
+   * This method should ONLY be called as a result of direct user action in the UI,
+   * with the intention of supporting legacy dapps that don't support EIP 1102.
+   *
+   * @param {string} origin - The origin to expose the account(s) to.
+   * @param {Array<string>} accounts - The account(s) to expose.
+   */
+  async legacyExposeAccounts (origin, accounts) {
+
+    const permissions = {
+      eth_accounts: {},
+    }
+
+    await this.finalizePermissionsRequest(permissions, accounts)
+
+    let error
+    try {
+      error = await new Promise((resolve) => {
+        this.permissions.grantNewPermissions(origin, permissions, {}, err => resolve(err))
+      })
+    } catch (err) {
+      error = err
+    }
+
+    if (error) {
+      if (error.code === 4001) {
+        throw error
+      } else {
+        throw ethErrors.rpc.internal({
+          message: `Failed to add 'eth_accounts' to '${origin}'.`,
+          data: {
+            originalError: error,
+            accounts,
+          },
+        })
+      }
+    }
+  }
+
+  /**
+   * Update the accounts exposed to the given origin.
+   * Throws error if the update fails.
+   *
+   * @param {string} origin - The origin to change the exposed accounts for.
+   * @param {string[]} accounts - The new account(s) to expose.
+   */
+  async updateExposedAccounts (origin, accounts) {
```

Comment:
This method is also unused, and would similarly be required to manually connect a pre-EIP-1102 dapp to an account (after it had already been connected to a different account).

Reply:
I don't think this is necessarily just for "manually connect a pre-EIP-1102 dapp to an account (after it had already been connected to a different account)". It could also be used for updating which accounts are connected to a given origin, regardless of method by which the existing connections were made.

We don't support this in the UI now, but I think we will in the future? So maybe it is okay to leave this as an addition to our exposed api that we will start consuming with some future feature.

Reply:
Ohh I see, that makes sense. 👍